drivers: serial: nrfx_uarte: UART shim based on nrfx_uarte driver

nordic-krch opened the pull request:
PR adds new Nordic UART shim which is based on the nrfx_uarte driver. Legacy shim is kept for now and can be used after `CONFIG_UART_NRFX_UARTE_LEGACY_SHIM=y`.

New shim implements only polling and asynchronous API. Interrupt driven API is implemented via generic asynchronous to interrupt driven adaptation layer. New shim is using different approach to RX bytes counting (legacy shim was using RXDRDY interrupt or dedicated TIMER peripheral).

New approach changes the behavior of the RX while being compliant with the API. When there is no RX activity for the specified amount of time UART RX is restarted (stopped with ENDRX_STARTRX short being set). With this approach there is no need for byte counting since DMA transfer ends on timeout and AMOUNT register is updated. However, it means that every `UART_RX_RDY` event comes with the new buffer (offset in `uart_rx_event` is always 0).

PR contains asynchronous to interrupt driven adaptation layer which is using `uart_async_rx` helper added by #63967.

PR is based on #65113 and #63967.

nordic-krch committed:
drivers: serial: Add async to interrupt driven adaptation

Add adaptation layer which allows to provide interrupt driven
API for drivers which exposes only asynchronous API.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: uart_mix_fifo_poll: Align the the new nrfx driver

Align test configurations to the new uart driver based on nrfx_uarte.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: arm: nordic_nrf: Disable UART runtime configuration

Since it takes 400 bytes of code and it is rarely used disable
by default this feature.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

dcpleung commented on drivers/serial/Kconfig:
```diff
@@ -133,6 +133,20 @@ config UART_ASYNC_RX_HELPER
 	  buffers.
 
+config UART_ASYNC_TO_INT_DRIVEN_API
+	bool "Asynchronous to Interrupt driven adaptation layer"
```
Maybe skip the prompt so only drivers needing this will enable this kconfig?

nordicjm reviewed (commented):
See comments in https://github.com/zephyrproject-rtos/zephyr/pull/67446

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+	}
+}
+
+static int uarte_nrfx_pm_action(const struct device *dev,
```
This function should be wrapped in `#ifdef CONFIG_PM_DEVICE`.

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+	}; \
+	static int CONCAT(uarte_, idx, _init)(const struct device *dev) \
+	{ \
+		(void)uarte_nrfx_pm_action;/* To avoid compilation warnings. */ \
```
Warnings about unused static functions?

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+
+
+
+#define UARTE_MEMORY_SECTION(idx)					       \
```
It would be better to locate this macro before `UART_NRF_UARTE_DEVICE`.

anangl reviewed (commented):
Just a few initial comments. I haven't managed to go through all the code yet.

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+			 (is_async_api(dev) ? NRFX_UARTE_RX_ENABLE_STOP_ON_END : 0);
+
+	if (timeout != SYS_FOREVER_US) {
+		/* Set minimum interval to 3 RTC ticks. 3 is used due to RTC limitation
+		 * which cannot set timeout for next tick. Assuming delay in processing
+		 * 3 instead of 2 is used. Note that lower value would work in a similar
+		 * way but timeouts would always occur later than expected,  most likely
+		 * after ~3 ticks.
+		 */
+		adata->idle_cnt = RX_TIMEOUT_DIV + 1;
+		adata->rx_timeout = K_USEC(timeout / RX_TIMEOUT_DIV);
+		adata->rx_timeout.ticks = MAX(3, adata->rx_timeout.ticks);
```
We got rid of such rounding up to 3 ticks in 52c55177ba71af328c6069ea56c007e6eb5cdf1d. Why is it coming back? See also https://github.com/nrfconnect/sdk-zephyr/pull/1336#discussion_r1360443426. Kernel ticks may not be equal to RTC ticks, and for GRTC the correction is completely inappropriate.

nordic-krch replied:
Removed.

anangl commented on include/zephyr/drivers/serial/uart_async_to_irq.h:
```diff
@@ -0,0 +1,267 @@
+
+/** @brief Data structure used by the adaptation layer.
+ *
+ * This data must be the first element of the UART device data structure.
```
A pointer to this data must be that first element. Same for the config.

nordic-krch replied:
Added. Also added static asserts in the shim to check that.

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+		return uarte_nrfx_init(dev); \
+	} \
+	PM_DEVICE_DT_DEFINE(UARTE(idx), uarte_nrfx_pm_action); \
+	IF_ENABLED(CONCAT(CONFIG_UART_, idx, _INTERRUPT_DRIVEN), \
+		(static int CONCAT(uart_rx_start, idx)(void) { \
+			static const struct device *dev = DEVICE_DT_GET(UARTE(idx)); \
+			start_rx(dev); \
+			return 0; \
+		} \
+		SYS_INIT(CONCAT(uart_rx_start, idx), POST_KERNEL, 0);)) \
```
Why is this done here, not in `uarte_nrfx_init()`?

nordic-krch replied:
Removed. I think that it got back to times then a2i adaptation layer was using some kernel calls and required that init function is called post kernel but it is not longer the case.

gmarull commented on include/zephyr/drivers/serial/uart_async_to_irq.h:
```diff
@@ -0,0 +1,267 @@
+ *
+ * This data must be the first element of the UART device data structure.
+ */
+struct uart_async_to_irq_data;
+struct uart_async_to_irq_config;
```
only one is documented

gmarull commented on include/zephyr/drivers/serial/uart_async_to_irq.h:
```diff
@@ -0,0 +1,267 @@
+/** @internal Interrupt driven interrupt update function */
+int z_uart_async_to_irq_irq_update(const struct device *dev);
+
+/** @internal Set the irq callback function */
+void z_uart_async_to_irq_irq_callback_set(const struct device *dev,
```
do not use `@internal`, guard all entries in `@cond INTERNAL_HIDDEN`

gmarull commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -26,10 +26,17 @@ config UART_NRFX_UARTE
 
+config UART_NRFX_UARTE_LEGACY_SHIM
+	depends on UART_NRFX_UARTE
+	# New shim takes more ROM. Until it is fixed use legacy shim in memory
+	# constraint case.
```
is this expected to improve/be fixed?

nordic-krch replied:
Yes, eventually we want to have one shim only. But harder to say exact when.

gmarull commented on drivers/serial/Kconfig.nrfx_uart_instance:
```diff
@@ -66,6 +73,43 @@ config UART_$(nrfx_uart_num)_NRF_ASYNC_LOW_POWER
 	depends on UART_$(nrfx_uart_num)_NRF_HW_ASYNC
+	depends on UART_NRFX_UARTE_LEGACY_SHIM
+
+config UART_$(nrfx_uart_num)_TX_CACHE_SIZE
+	int "TX cache buffer size"
+	depends on !UART_NRFX_UARTE_LEGACY_SHIM
+	default 8
+	help
+	  For UARTE, TX cache buffer is used when provided TX buffer is not located
+	  in memory which can be used by the EasyDMA.
+
+config UART_$(nrfx_uart_num)_RX_CACHE_SIZE
+	int "RX cache buffer size"
+	depends on !UART_NRFX_UARTE_LEGACY_SHIM
+	default 32 if $(dt_nodelabel_has_compat,ram3x,$(DT_COMPAT_MMIO_SRAM))
+	default 5
+	range 5 255
+	help
+	  For UARTE, RX cache buffer is used when provided RX buffer is not located
+	  in memory which can be used by the EasyDMA. It is also used to store
+	  flushed data.
+
+config UART_$(nrfx_uart_num)_A2I_RX_SIZE
+	depends on !UART_NRFX_UARTE_LEGACY_SHIM
+	int "Asynchronous to interrupt driven adaptation layer RX buffer size"
+	default 64 if UART_$(nrfx_uart_num)_INTERRUPT_DRIVEN
+	default 0
+	help
+	  Amount of space dedicated for RX. It is divided into chunks with some
+	  amount of that space used for control data.
+
+config UART_$(nrfx_uart_num)_A2I_RX_BUF_COUNT
```
not a fan is instance-based options, these do not scale well and perpetuate the problem nrfx-based driver have: hardcoded relationship with hw instance numbers. Isn't there a better option?

gmarull replied:
ping

gmarull commented on modules/hal_nordic/nrfx/Kconfig:
```diff
@@ -573,6 +573,26 @@ config NRFX_UARTE3
 	depends on $(dt_nodelabel_has_compat,uart3,$(DT_COMPAT_NORDIC_NRF_UARTE))
 	select NRFX_UARTE
 
+config NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG
+	bool "UARTE GPIO configuration support"
+	default y
+
+config NRFX_UARTE_CONFIG_SKIP_PSEL_CONFIG
+	bool "UARTE PSEL configuration support"
+	default y
```
why is this required? doesn;t driver use pinctrl unconditionally?

nordic-krch replied:
we want to tell nrfx to not compile the support for handling GPIO and PSEL in the driver. I changed to so that uart shim is actually implying those options (and removed defaults from here).

gmarull commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+ * @brief Driver for Nordic Semiconductor nRF UARTE
+ */
+
+#include <zephyr/drivers/uart.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+#include <haly/nrfy_uarte.h>
+#include <helpers/nrfx_gppi.h>
+#include <nrfx_uarte.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
+#include <zephyr/linker/devicetree_regions.h>
+#include <zephyr/irq.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/serial/uart_async_to_irq.h>
+#include <haly/nrfy_gpio.h>
```
please sort headers, local, zephyr, hal/3rd party

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -26,10 +26,17 @@ config UART_NRFX_UARTE
+	# New shim takes more ROM. Until it is fixed use legacy shim in memory
+	# constraint case.
+	default y if MCUBOOT
+	bool "Legacy UARTE shim"
```
It's better to place the option type and prompt on top, for consistency.

nordic-krch commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+#define UARTE_CFG_FLAG_PPI_ENDTX   BIT(1)
+
+/* If set then receiver is not used. */
+#define UARTE_CFG_FLAG_NO_RX BIT(3)
```
It's no used after update to use `disable-rx` DT flag.

nordic-krch commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+
+	if ((IS_ENABLED(CONFIG_UART_USE_RUNTIME_CONFIGURE) &&
+	     (data->uart_config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS)) ||
+	     (config->config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS)) {
```
Reworked it so that `uart_config` is in data structure only if runtime configuration is enabled. Otherwise, hwfc is checked in nrfx config structure.

nordic-krch commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+			.skip_gpio_cfg = true, \
+		}, \
+		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(UARTE(idx)), \
+		.config = { \
```
Removed this structure as it was redundant.

nordic-krch commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1058 @@
+	nrfx_uarte_t nrfx_dev;
+	nrfx_uarte_config_t nrfx_config;
+	const struct pinctrl_dev_config *pcfg;
+	struct uart_config config;
```
Removed `uart_config` structure for driver configuration structure.

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1053 @@
+		 * general is not power efficient and should be avoided in low
+		 * power applications.
+		 */
+		err = nrfx_uarte_tx(nrfx_dev, &out_char, 1, NRFX_UARTE_TX_EARLY_RETURN);
+		__ASSERT(err != NRFX_ERROR_INVALID_ADDR, "Invalid address of the buffer");
```
Shouldn't there be something like `if (err != NRFX_ERROR_BUSY) break;` here to avoid potential hanging when assertions are disabled or when an unexpected error (like `NRFX_ERROR_FORBIDDEN`) occurs?

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1053 @@
+	return 0;
+}
+
+static int start_rx(const struct device *dev)
+{
+	const struct uarte_nrfx_config *cfg = dev->config;
+
+	if (cfg->flags & UARTE_CFG_FLAG_NO_RX) {
+		return -ENOTSUP;
```
`start_rx()` is called at initialization of the device, so this will cause the initialization to fail for instances not using the async API but having the `disable-rx` property set.

nordic-krch replied:
Added error code conversion from -ENOTSUP to 0 in the init function so that -ENOTSUP is also accepted.

anangl commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -175,6 +175,9 @@ Device Drivers and Device Tree
                      <&rcc STM32_SRC_PLL1_Q FDCAN_SEL(1)>;
     };
 
+* Runtime configuration of UART driver is now by default disabled for Nordic UART driver. Motivation
+  for that change is that feature is rarely used and disabling reduces memory footprint.
```
I'd rephrase this a bit. For example: "Runtime configuration is now disabled by default for Nordic UART drivers. The motivation for the change is that this feature is rarely used and disabling it significantly reduces the memory footprint."

nordic-krch commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1053 @@
+			evt.data.rx_stop.reason = UARTE_ERROR_FROM_MASK(data->async->err);
+			evt.data.rx_stop.data.buf = event->data.rx.p_buffer;
+			evt.data.rx_stop.data.len = event->data.rx.length;
+			if (!(config->flags & UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API)) {
```
Actually I had to change those static inline to macros. That is because not all compilers are smart enough to determine from static inline that there is a dead code. There were linker issues when compiling for nrf52_bsim.

nordic-krch commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1053 @@
+		adata->rx_timeout = K_NO_WAIT;
+	}
+
+	atomic_or(&data->flags, UARTE_DATA_FLAG_RX_ENABLED);
```
moved down and setting only on success.

nordic-krch commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1053 @@
+					  async_to_irq_trampoline,				\
+					  UARTE_PROP(idx, current_speed),			\
+					  uarte##idx##_tx_cache,				\
+					  CONFIG_UART_##idx##_TX_CACHE_SIZE - 1,		\
```
Yes, nrfx_uarte when it's getting a TX cache buffer, it is using the last byte in the cache buffer for poll_out so it cannot be used by the a2i layer.

anangl commented on modules/hal_nordic/nrfx/Kconfig:
```diff
@@ -573,6 +573,24 @@ config NRFX_UARTE3
 	depends on $(dt_nodelabel_has_compat,uart3,$(DT_COMPAT_NORDIC_NRF_UARTE))
 	select NRFX_UARTE
 
+config NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG
+	bool "UARTE GPIO configuration support"
+
+config NRFX_UARTE_CONFIG_SKIP_PSEL_CONFIG
+	bool "UARTE PSEL configuration support"
+
+config NRFX_UARTE_CONFIG_TX_LINK
+	bool "UARTE TX transfer linking support"
+
+config NRFX_UARTE_CONFIG_RX_CACHE_ENABLED
```
All these options should have `depends on NRFX_UARTE`.

anangl commented on modules/hal_nordic/nrfx/Kconfig:
```diff
@@ -573,6 +573,24 @@ config NRFX_UARTE3
+
+config NRFX_UARTE_CONFIG_RX_CACHE_ENABLED
+	bool "UARTE RX caching support"
+	depends on $(dt_nodelabel_has_compat,ram3x,$(DT_COMPAT_MMIO_SRAM))
+	default y
```
Shouldn't this be rather:
```suggestion
	default y if $(dt_nodelabel_has_compat,ram3x,$(DT_COMPAT_MMIO_SRAM))
```
?
Is there a reason to disallow this option when there is no such node?

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1032 @@
+/* Determine if the device has interrupt driven API enabled. */
+#define IS_INT_DRIVEN_API(dev) \
+	(IS_ENABLED(CONFIG_UART_ASYNC_TO_INT_DRIVEN_API) && \
+	 (((const struct uarte_nrfx_config *)dev->config)->flags & \
+	 UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API))
```
```suggestion
	 (((const struct uarte_nrfx_config *)dev->config)->flags & \
	  UARTE_CFG_FLAG_INTERRUPT_DRIVEN_API))
```

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1032 @@
+
+/* Determine if the device has interrupt driven API enabled. */
+#define IS_INT_DRIVEN_API(dev) \
+	(IS_ENABLED(CONFIG_UART_ASYNC_TO_INT_DRIVEN_API) && \
```
In theory, `CONFIG_UART_ASYNC_TO_INT_DRIVEN_API` could be enabled for some other UART driver.
```suggestion
	(UARTE_ANY_INTERRUPT_DRIVEN && \
```

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1032 @@
+	const struct uarte_nrfx_config *cfg = dev->config;
+
+	if (cfg->flags & UARTE_CFG_FLAG_NO_RX) {
+		return -ENOTSUP;
```
This will make `PM_DEVICE_ACTION_SUSPEND` fail. `stop_rx()` should not be called when RX is disabled as `start_rx()` is not called then.

anangl reviewed (commented):
LGTM. Just a few nitpicking comments, except for the last one.

anangl commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -564,6 +564,22 @@
+#ifdef CONFIG_NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG
+#define NRFX_UARTE_CONFIG_SKIP_GPIO_CONFIG 1
+#endif
+
```
These blank lines are inconsistent with what we have for other drivers.
And actually these translations and the Kconfig options themselves should be introduced in a separate commit.

nordic-krch committed:
modules: hal_nordic: nrfx: Add new UARTE options to Kconfig

Add Kconfig options to new configuration flags for nrfx_uarte.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx: Add new shim based on nrfx_uarte

Add new shim which is based on nrfx driver.

Legacy shim is kept for transition period. It can be used after
setting CONFIG_UART_NRFX_UARTE_LEGACY_SHIM.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

anangl commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1031 @@
+
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		if ((IS_INT_DRIVEN_API(dev) || IS_POLLING_API(dev)) &&
+		     !(cfg->flags & UARTE_CFG_FLAG_NO_RX)) {
```
For consistency with `PM_DEVICE_ACTION_RESUME`:
```suggestion
		if (!IS_ASYNC_API(dev) && !(cfg->flags & UARTE_CFG_FLAG_NO_RX)) {
```

anangl commented on drivers/serial/uart_async_to_irq.c:
```diff
@@ -0,0 +1,377 @@
+
+static const struct uart_async_to_irq_config *get_config(const struct device *dev)
+{
+	const struct uart_async_to_irq_config **config = dev->config;
```
Sorry, I missed that one `const` in my previous comment.
```suggestion
	const struct uart_async_to_irq_config *const *config = dev->config;
```

nordic-krch committed:
tests: drivers: uart: async_api: Add configuration for legacy nrf driver

Add configuration which tests legacy nrf uart driver.
Prevent tests on nrf52_bsim of the new driver (it is requiring
adjustments for the simulation). It will be fixed by another
patch.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: uart: uart_pm: Use legacy nrf driver

New nrf uart shim requires addiotional work around PM. Use legacy
version for those tests until it is fixed.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
boards: posix: nrf_bsim: Use legacy UART shim

New UART shim must be adapted to work in the simulated environment.
Use legacy version until it is fixed.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

gmarull reviewed (dismissed):
IMHO this should be fixed better https://github.com/zephyrproject-rtos/zephyr/pull/65155/files#r1450353386, but let's move on.

aescolar commented on drivers/serial/uart_nrfx_uarte2.c:
```diff
@@ -0,0 +1,1030 @@
+			k_msleep(1);
+		}
+		Z_SPIN_DELAY(3);
+	} while (err != NRFX_ERROR_BUSY);
```
Isn't this a bug? shouldn't it be 
`} while (err != NRFX_SUCCESS);` ?
Otherwise the same byte would be enqueued repeatedly if the Tx is enqueued right away, or?

anangl replied:
Yes, it should be `while (err == NRFX_ERROR_BUSY)`.

anangl reviewed (changes requested):
See https://github.com/zephyrproject-rtos/zephyr/pull/65155#discussion_r1456033866.
